Source/Dafny/Dafny.atg

```diff
@@ -4105,13 +4133,22 @@ CasePatternLocal<.out CasePattern<LocalVariable> pat, bool isGhost.>
     ")"                                          (. pat = new CasePattern<LocalVariable>(id, id.val, arguments); .)
   | "("                                          (. id = t;
                                                     arguments = new List<CasePattern<LocalVariable>>();
+                                                    ghostArguments = new List<bool>();
                                                  .)
-      [ CasePatternLocal<out pat, isGhost>       (. arguments.Add(pat); .)
-        { "," CasePatternLocal<out pat, isGhost> (. arguments.Add(pat); .)
+      [
+        [ "ghost"                                        (. ghostArg = true; .)
+        ]
+        CasePatternLocal<out pat, isGhost || ghostArg>   (. arguments.Add(pat); ghostArguments.Add(isGhost || ghostArg); .)
+        { ","                                            (. ghostArg = false; .)
+          [ "ghost"                                      (. ghostArg = true; .)
+          ]
+          CasePatternLocal<out pat, isGhost || ghostArg> (. arguments.Add(pat); ghostArguments.Add(isGhost || ghostArg); .)
         }
       ]
     ")"                                (. // Parse parenthesis without an identifier as a built in tuple type.
-                                          theBuiltIns.TupleType(id, arguments.Count, true); // make sure the tuple type exists
+                                          // TODO(fpoli): Why is the following call needed?
+                                          // TODO(fpoli): Given the statement`ghost var (x, ghost y) := ...;`, should we consider the type of the tuple to be `(ghost, ghost)` or `(non-ghost, ghost)`?
```

Comment:
I would make this illegal since adding a redundant `ghost` implies the programmer didn't understand what they were doing.

Reply:
It sounds good to me. I can easily generate a `SemErr(..)` for those cases.

Reply:
I thought about this some more. Now, I don't think it's all that tricky. In fact, I don't think there's any difference from how things work today. Today, the `match` source expression is type checked. Then, the type inference machinery makes an effort to figure out the type of this expression (which it does by calling `PartiallySolveTypeConstraints()` at that time). If it figures out the type and it's some kind of tuple type, then it known what type to look up the constructors in. If it can't figure it out, it gives an error. So, I expect that we'll always know what type to look up the constructor in. (We may need to add a call to `PartiallySolveTypeConstraints()` in the resolution of `var (x, y) := e;` statements and `var (x, y) := e0; e1` expressions. I don't know if those calls are there today.)

Reply:
`ghost var (x, ghost y) := ...;` is now a syntax error, as the inner ghost is not allowed by the grammar.